plugins/jobtype/src/azkaban/jobtype/HadoopSecureSparkWrapper.java

```diff
@@ -62,47 +86,142 @@ public static void main(final String[] args) throws Exception {
+        && argArray[i+1].startsWith(SPARK_CONF_EXTRA_DRIVER_OPTIONS)) {
+        driverJavaOptions.append(argArray[++i].substring(SPARK_CONF_EXTRA_DRIVER_OPTIONS.length() + 1));
+      }
+    }
     
-    StringBuilder driverJavaOptions = new StringBuilder(newArgs[1]);    
+    // Append addtional driver java opts about azkaban context
     String[] requiredJavaOpts = { WORKFLOW_LINK, JOB_LINK, EXECUTION_LINK, ATTEMPT_LINK };
     for (int i = 0; i < requiredJavaOpts.length; i++) {
-    	  driverJavaOptions.append(" ").append(HadoopJobUtils.javaOptStringFromHadoopConfiguration(conf,
-    	            requiredJavaOpts[i]));
+        driverJavaOptions.append(" ").append(HadoopJobUtils.javaOptStringFromHadoopConfiguration(conf,
+                  requiredJavaOpts[i]));
     }
-    newArgs[1] = driverJavaOptions.toString();
-    logger.info("newArgs2: " + Arrays.toString(newArgs));
+    // Update driver java opts
+    argArray[1] = driverJavaOptions.toString();
+  }
 
-    org.apache.spark.deploy.SparkSubmit$.MODULE$.main(newArgs);
+  private static void handleDynamicResourceAllocation(String[] argArray) {
+    // HadoopSparkJob will set env var on this process if we enable dynamic allocation policy for spark jobtype.
+    // This policy can be enabled in spark jobtype plugin's conf property.
+    // This policy is designed to enforce dynamic allocation.
+    String dynamicAllocProp = System.getenv(HadoopSparkJob.SPARK_DYNAMIC_RES_ENV_VAR);
+    boolean dynamicAllocEnabled = dynamicAllocProp != null && dynamicAllocProp.equals(Boolean.TRUE.toString());
+    if (dynamicAllocEnabled) {
+      for (int i = 0; i < argArray.length; i++) {
+        if (argArray[i] == null) continue;
+
+        // If user specifies num of executors, or if user tries to disable dynamic allocation for his application
+        // by setting some conf params to falase, we need to ignore these settings to enforce the application 
+        // uses dynamic allocation.
+        if (argArray[i].equals(SparkJobArg.NUM_EXECUTORS.sparkParamName) // --num-executors
+          || (
+            argArray[i].equals(SparkJobArg.SPARK_CONF_PREFIX.sparkParamName) // --conf
+            && (argArray[i+1].startsWith(SPARK_CONF_SHUFFLE_SERVICE_ENABLED) // spark.shuffle.service.enabled=
+              || argArray[i+1].startsWith(SPARK_CONF_DYNAMIC_ALLOC_ENABLED)) // spark.dynamicAllocation.enabled
+        )) {
+
+          logger.info("Azbakan enforces dynamic resource allocation. Ignore user param: "
+            + argArray[i] + " " + argArray[i+1]);
+          argArray[i] = null;
+          argArray[++i] = null;
+        }
+      }
+    }
+  }
+
+  private static void handleNodeLabeling(String[] argArray) {
+    // HadoopSparkJob will set env var on this process if we enable node labeling policy for spark jobtype.
+    // We also detect the yarn cluster settings has enable node labeling
+    // We enforce node labeling policy by ignoring user queue params.
```

Comment:
Better to also clearly point out the differences between enabling node labeling on YARN side vs. setting node labeling policy on the Spark job type side.